samples/azure-quantum/shorter-quantum-circuits/shorter-quantum-circuits-dataset.ipynb

```diff
@@ -0,0 +1,782 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import matplotlib.pyplot as plt\n",
+    "import pandas\n",
+    "import math\n",
+    "import data_analysis\n",
+    "\n",
+    "res = []\n",
+    "cost_labels = {\"rcount\":\"non-Clifford rotation count\",\"tcost\" : \"T-count\", \"power\" : \"determinant power\" }\n",
+    "files_limit = 8 # change to None to process the datasets completely, it will take a long time\n",
+    "for dataset_file in [\"clifford-t-random-angles.tar.gz\",\"clifford-root-t-random-angles.tar.gz\",\"clifford-t-fourier-angles.tar.gz\",\"clifford-root-t-fourier-angles.tar.gz\"]:\n",
+    "    gate_costs_of = data_analysis.dataset_costs[dataset_file]\n",
+    "    if dataset_file.endswith('random-angles.tar.gz'):\n",
+    "        for cost in gate_costs_of:\n",
+    "            for protocol in [ \"diagonal\", \"fallback\", \"mixed diagonal\", \"mixed fallback\" ] :\n",
+    "                name_with_dashes = protocol.replace(' ', '-')\n",
+    "                protocol_data = pandas.read_csv(f'./data/{dataset_file.split(\".\")[0]}-{name_with_dashes}.tsv', sep = \"\\t\", header=0, index_col=0)\n",
+    "                max_count = max([x for x in protocol_data[\"count\"]])\n",
+    "                filtered_data = protocol_data.query(f\"error < {1e-2} & count == {max_count}\")\n",
+    "                param, cov = data_analysis.best_fit_parameters_of(filtered_data, cost_column = cost)\n",
+    "                param_max, cov_max = data_analysis.best_fit_parameters_of(filtered_data, cost_column = cost + \":max\", use_sigmas=False)\n",
+    "                param_min, cov_min = data_analysis.best_fit_parameters_of(filtered_data, cost_column = cost + \":min\", use_sigmas=False)\n",
+    "                param_median, cov_median = data_analysis.best_fit_parameters_of(filtered_data, cost_column = cost + \":median\", use_sigmas=False)\n",
+    "                dataset_name = dataset_file.split(\".\")[0]\n",
+    "                res.append({\"dataset (cost)\":f\"{dataset_name} ({cost})\",\n",
+    "                    \"protocol\":protocol,\n",
+    "                    \"mean cost fit\":f\"{param[0]:.2f}log₂(1/ε)+{param[1]:.2f}\",\n",
+    "                    \"mean cost fit stddev\":f\"({cov[0,0]:.2e}, {cov[1,1]:.2e})\",\n",
+    "                    \"max cost fit\":f\"{param_max[0]:.2f}log₂(1/ε)+{param_max[1]:.2f}\",\n",
+    "                    \"max cost fit stddev\":f\"({cov_max[0,0]:.2e}, {cov_max[1,1]:.2e})\",\n",
+    "                    \"min cost fit\":f\"{param_min[0]:.2f}log₂(1/ε)+({param_min[1]:.2f})\",\n",
+    "                    \"median cost fit\":f\"{param_median[0]:.2f}log₂(1/ε)+{param_median[1]:.2f}\",\n",
+    "                    })\n",
+    "\n",
+    "display(pandas.DataFrame(res).groupby(\"dataset (cost)\",group_keys=True).apply(lambda a: a.drop([\"dataset (cost)\"], axis=1)[:]))               \n",
+    "            "
+   ]
+  },
+  {
+   "cell_type": "code",
```

Comment:
Try not to finish with an empty code cell. It would be better to have some concluding text cell.

Reply:
Removed cell